### ISSUE TITLE
Typing is wrong for overload of __round__ when ndigits parameter is omitted

### ISSUE THREAD

**Issue by migwellian** [python/typeshed#2638]
example code:

	from typing import SupportsRound

	class MyClass(SupportsRound['MyClass']):

	    def __round__(self: 'MyClass', ndigits: int = 0) -> 'MyClass':
	        return self

	def test_tmp() -> None:
	    x = MyClass()
	    result: MyClass

	    result = round(x, 0)
	    assert type(result) == MyClass
	    result = round(x)
	    assert type(result) == MyClass

Output from MyPy:

> 16: error: Incompatible types in assignment (expression has type "int", variable has type "MyClass")

Note that there is no error on line 14 (i.e. the type is correctly inferred when I pass in a value for `ndigits`), and that the test passes (i.e. in both calls to `round` I get back an object of type `MyClass`)

Version numbers: Python 3.6.5, mypy 0.641.

**Comment by Michael0x2a** [comment]
Background context: this question [was originally asked on StackOverflow](https://stackoverflow.com/q/53481887/646543). I came to the conclusion there that the definition of `SupportRound` might be inconsistent with the definition of `round` and should be changed, but told this user to try asking here in case there was some nuance I'm overlooking.

**Comment by srittau** [comment]
I feel that `round()` called on a `SupportsRound[T]` should always return `T`. [The documentation](https://docs.python.org/3/library/functions.html#round) is not really clear on what happens when delegating to `__round__()`, but I'd be surprised if the average implementations returns a different type for when `ndigits` is supplied.

Another (hard to implement, but probably the correct) solution is to give `SupportsRound` a second, optional type argument.

**Comment by hauntsaninja** [comment]
Based on https://docs.python.org/3/library/functions.html#round, it seems that if you don't specify `ndigits`, `round` will return an `int`: 
> If ndigits is omitted or is None, it returns the nearest integer to its input.

Note that this is, in fact, the behaviour for `float`:
```
~ λ python3
Python 3.7.5 (default, Nov  1 2019, 02:16:23) 
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> round(3.2, 0)
3.0
>>> round(3.2)
3
```

Giving `SupportsRound` a second, optional type parameter could work.
This could maybe help clarify cases like https://github.com/python/mypy/issues/8213 where the current type parameter isn't needed by being more explicit (e.g, maybe that issue would be less of a stumbling block if `SupportsRound[Any, int]` was a thing).

**Comment by hauntsaninja** [comment]
Note that #2907 adds an additional error message to the original case, as Michael suggested, which will help force the user to handle both cases when defining MyClass.
`5: error: Signature of "__round__" incompatible with supertype "SupportsRound"`

**Comment by Avasam** [comment]
This seems to have been solved. This example all passes:
```python
from typing import SupportsRound
from typing_extensions import Self, assert_type


class MyClass(SupportsRound["MyClass"]):

    def __round__(self, ndigits: int = 0) -> Self:
        return self


def test_tmp() -> None:
    x = MyClass()

    assert_type(round(x, 0), MyClass)
    assert_type(round(x), MyClass)
    assert_type(round(0.0, 0), float)
    assert_type(round(0.0), int)
```

**Comment by AlexWaygood** [comment]
> This seems to have been solved.

Nice! Probably fixed by #9151
